homeassistant/components/zha/core/device.py

```diff
@@ -0,0 +1,319 @@
+    def last_seen(self):
+        """Return last_seen for device."""
+        return self._zigpy_device.last_seen
+
+    @property
+    def manufacturer_code(self):
+        """Return manufacturer code for device."""
+        # will eventually get this directly from Zigpy
+        return None
+
+    @property
+    def gateway(self):
+        """Return the gateway for this device."""
+        return self._zha_gateway
+
+    @property
+    def cluster_listeners(self):
+        """Return cluster listeners for device."""
+        return self._cluster_listeners.values()
+
+    @property
+    def all_listeners(self):
+        """Return cluster listeners and relay listeners for device."""
+        return self._all_listeners
+
+    @property
+    def cluster_listener_keys(self):
+        """Return cluster listeners for device."""
+        return self._cluster_listeners.keys()
+
+    @property
+    def available_signal(self):
+        """Signal to use to subscribe to device availability changes."""
+        return self._available_signal
+
+    @property
+    def available(self):
+        """Return True if sensor is available."""
+        return self._available
+
+    def update_available(self, available):
+        """Set sensor availability."""
+        if self._available != available and available:
+            # Update the state the first time the device comes online
+            async_dispatcher_send(
+                self.hass,
+                self._available_signal,
+                False
+            )
+            async_dispatcher_send(
+                self.hass,
+                "{}_{}".format(self._available_signal, 'entity'),
+                True
+            )
+        self._available = available
+
+    @property
+    def device_info(self):
+        """Return a device description for device."""
+        ieee = str(self.ieee)
+        return {
+            IEEE: ieee,
+            ATTR_MANUFACTURER: self.manufacturer,
+            MODEL: self.model,
+            NAME: self.name or ieee
+        }
+
+    def add_cluster_listener(self, cluster_listener):
+        """Add cluster listener to device."""
+        # only keep 1 power listener
+        if cluster_listener.name is LISTENER_BATTERY and \
+                LISTENER_BATTERY in self._cluster_listeners:
+            return
+        self._all_listeners.append(cluster_listener)
+        if isinstance(cluster_listener, EventRelayListener):
+            self._relay_listeners.append(cluster_listener)
+        else:
+            self._cluster_listeners[cluster_listener.name] = cluster_listener
+
+    def get_cluster_listener(self, name):
+        """Get cluster listener by name."""
+        return self._cluster_listeners.get(name, None)
+
+    async def async_configure(self):
+        """Configure the device."""
+        _LOGGER.debug('%s: started configuration', self.name)
+        for listener in self.all_listeners:
+            try:
+                await listener.async_configure()
```

Comment:
Does the order matter here, or could we let asyncio handle all the tasks, via eg asyncio.wait or asyncio.gather? What do we want to happen on exception? Should we stop all on first exception, or try to finish, etc?

Reply:
order of individual listeners doesn't matter. we can let asyncio handle so long as we can await the execution of the entire function call. exceptions should be treated as warnings meaning we should do as much as we can so the device function degrades as little as possible.

Reply:
@MartinHjelmare is this kind of what you were going for:

```
    async def async_configure(self):
        """Configure the device."""
        _LOGGER.debug('%s: started configuration', self.name)
        listener_tasks = []
        for listener in self.all_listeners:
            listener_tasks.append(
                self._async_configure_listener(listener.async_configure()))
        await asyncio.gather(*listener_tasks)
        _LOGGER.debug('%s: completed configuration', self.name)


    async def _async_configure_listener(self, listener):
        """Configure a single listener on this device."""
        try:
            await listener.async_configure()
            _LOGGER.debug('%s: listener: %s has been configured',
                          self.name,
                          "{}-{}".format(
                              listener.name, listener.unique_id))
        except Exception as ex:  # pylint: disable=broad-except
            _LOGGER.warning(
                '%s errored while configuring listener: %s ex: %s',
                self.name,
                "{}-{}".format(listener.name, listener.unique_id),
                ex
            )
```

Reply:
Looks good!

Reply:
I’ll push that change in a few